### PR TITLE
Mednafen refactors

### PR DESCRIPTION
**Pull request by AndersonTorres**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/pull/304028

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/me/mednaffe/package.nix [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  autoreconfHook,
+  fetchFromGitHub,
+  gtk3,
+  mednafen,
+  pkg-config,
+  stdenv,
+  wrapGAppsHook,
+}:
+
+stdenv.mkDerivation (finalAttrs: {
```
Comment (RossComputerGuy) [inline pkgs/by-name/me/mednaffe/package.nix:12]: Why `(finalAttrs: {` instead of just `rec {`?

File: pkgs/by-name/me/mednafen-server/package.nix [hunk 2]
```diff
@@ -0,0 +1,30 @@
+{
+  lib,
+  stdenv,
+  fetchurl,
+}:
+
+stdenv.mkDerivation (finalAttrs: {
```
Comment (RossComputerGuy) [inline pkgs/by-name/me/mednafen-server/package.nix:7]: Same here. Why `(finalAttrs: {` instead of just `rec {`?

File: pkgs/by-name/me/mednafen-server/package.nix [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    hash = "sha256-uJmxaMW+bydfAXq8XDOioMoBOLUsi5OT2Tpbbotsp3Y=";
+  };
+
+  strictDeps = true;
+
+  postInstall = ''
+    install -m 644 -Dt $out/share/mednafen-server standard.conf
+  '';
+
+  meta = {
```
Comment (RossComputerGuy) [inline pkgs/by-name/me/mednafen-server/package.nix:22]: If you make `meta = with lib; {` then you can remove `lib.` from anything below this line.

File: pkgs/by-name/me/mednafen/package.nix [hunk 4]
```diff
@@ -1,33 +1,39 @@
...
+  libjack2,
+  libsamplerate,
+  libsndfile,
+  pkg-config,
+  stdenv,
+  zlib,
 }:
 
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: {
```
Comment (RossComputerGuy) [inline pkgs/by-name/me/mednafen/package.nix:22]: It was just fine with `rec` before. Why the change?

File: pkgs/by-name/me/mednafen/package.nix [hunk 5]
```diff
@@ -1,33 +1,39 @@
...
-    url = "https://mednafen.github.io/releases/files/${pname}-${version}.tar.xz";
+    url = "https://mednafen.github.io/releases/files/mednafen-${finalAttrs.version}.tar.xz";
     hash = "sha256-2j+88Ch3+b4PAov6XRy1npU6QEm5D+fjk4ijOG2fNi4=";
   };
 
-  nativeBuildInputs = [ pkg-config ];
+  outputs = [ "out" "doc" ];
+
+  nativeBuildInputs = [
+    SDL2
```
Comment (RossComputerGuy) [inline pkgs/by-name/me/mednafen/package.nix:34]: Why is `SDL2` in `nativeBuildInputs` if it is in `buildInputs`?

File: pkgs/by-name/me/mednafen/package.nix [hunk 6]
```diff
@@ -48,16 +54,21 @@ stdenv.mkDerivation rec {
...
+
   postInstall = ''
-    mkdir -p $out/share/doc
-    mv Documentation $out/share/doc/mednafen
+    mkdir -p $doc/share/doc
+    mv Documentation $doc/share/doc/mednafen
   '';
 
-  meta = with lib; {
+  meta = {
```
Comment (RossComputerGuy) [inline pkgs/by-name/me/mednafen/package.nix:71]: Why the removal of `with lib;`?

**Review by RossComputerGuy — COMMENTED** [review]
Good start, though needs some improvements. Another thing to note, please stick with the [git commit naming conventions we use](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions). You can use `git rebase -i` to squash and rename commits so the history looks cleaner and you can follow the convention.

File: pkgs/by-name/me/mednafen/package.nix (hunk 5 above)
Comment (AndersonTorres) [reply pkgs/by-name/me/mednafen/package.nix:34]: When using `strictDeps = true`, we need two SDLs: the one that runs `sdl-config` from building platform and the other that installs SDL libs to target platform.

https://discourse.nixos.org/t/always-be-cross-compiling-slides-by-matthew-bauer-pdf/30287

File: pkgs/by-name/me/mednaffe/package.nix (hunk 1 above)
Comment (RossComputerGuy) [reply pkgs/by-name/me/mednaffe/package.nix:12]: Because the majority of other packages use rec and achieve the same result as what wrapping the attributes in a function does.

File: pkgs/by-name/me/mednafen-server/package.nix (hunk 3 above)
Comment (AndersonTorres) [reply pkgs/by-name/me/mednafen-server/package.nix:22]: `with` does not work like the distributive axiom of a field (as suggested by your remark), and its rules are unintuitive.

https://nix.dev/anti-patterns/language#with-attrset-expression

https://github.com/NixOS/nix/issues/490

https://github.com/NixOS/nixpkgs/issues/208242

File: pkgs/by-name/me/mednaffe/package.nix (hunk 1 above)
Comment (AndersonTorres) [reply pkgs/by-name/me/mednaffe/package.nix:12]: > Because the majority of other packages use rec

Popularity does not make a good argument here, or else things like treewide banishing unquoted URLs from the codebase would never happen.

> and achieve the same result as what wrapping the attributes in a function does

Although being spread over the whole Nixpkgs codebase, `rec` is an anti-pattern:

https://nix.dev/guides/best-practices#recursive-attribute-set-rec

Then, `finalAttrs` design pattern (a.k.a. __overlay style overridable recursive attributes__) was merged more than 3 years ago, to overcome some of the problems of `rec`.

> Also, what does this change fix or improve that required the removal of rec and wrapping it in a function?

#119942

Comment (RossComputerGuy) [reply pkgs/by-name/me/mednaffe/package.nix:12]: > Popularity does not make a good argument here, or else things like treewide banishing unquoted URLs from the codebase would never happen.

I am aware, however I believe the change should be made with a purpose.

> Although being spread over the whole Nixpkgs codebase, rec is an anti-pattern:
>
> https://nix.dev/guides/best-practices#recursive-attribute-set-rec
>
> Then, `finalAttrs` design pattern (a.k.a. overlay style overridable recursive attributes) was merged more than 3 years ago, to overcome some of the problems of `rec`.

Is there an RFC or issue about treewide removal of `rec`? The way it is upstream works fine currently. That nix.dev page recommends using `let...in` as an option.

File: pkgs/by-name/me/mednafen-server/package.nix (hunk 3 above)
Comment (RossComputerGuy) [reply pkgs/by-name/me/mednafen-server/package.nix:22]: That last one is about overuse of `with`. In this case, I wouldn't consider using `with lib;` for assigning values in `meta` to be an overuse.

File: pkgs/by-name/me/mednafen/package.nix (hunk 5 above)
Comment (AndersonTorres) [reply pkgs/by-name/me/mednafen/package.nix:34]: How exactly?

If in the sense of "it works on my machine, so it is not needed", then it is not needed.

If in the sense of "bad with it, worse without it", it is seriously needed.

Among many things, `strictDeps` enforces a cleaner separation of rules and concerns among inputs and phases - a thing that is easily blurred in native compilation but can be very tricky in cross compilation.

https://nixos.org/manual/nixpkgs/stable/#ssec-stdenv-dependencies-reference

https://github.com/NixOS/nixpkgs/issues/178468

Comment (RossComputerGuy) [reply pkgs/by-name/me/mednafen/package.nix:34]: Understandable, I wasn't aware that this was an issue with `nativeBuildInputs` and `buildInputs`.

File: pkgs/by-name/me/mednaffe/package.nix (hunk 1 above)
Comment (AndersonTorres) [reply pkgs/by-name/me/mednaffe/package.nix:12]: > I am aware, however I believe the change should be made with a purpose.

"Getting rid of `rec` (and other bad design patterns)" is a purpose, I suppose.

> Is there an RFC or issue about treewide removal of rec?

Reasoning backwards it is.

First, technically there is no need of a "treewide-rec-banish RFC" to banish `rec` from Nixpkgs.

Second, refactoring the expressions I maintain is in my discretion.

Third, objectively the code is not worse than before; it is still perfectly readable and understandable.

Fourth, and related to second, subjectively the code is better since it avoids pitfalls from bad design patterns, and it is better to me to maintain and improve it.

> The way it is upstream works fine currently.

Not a sufficient reason to hinder code refactoring.

> That nix.dev page recommends using let...in as an option.

Yes, and I do use it sometimes. 

But this nix.dev tutorial recommends let-in because, among other things, its purpose is to use more general approaches, especially in external codebases that will not necessarily have direct access to Nixpkgs.

Also, things like #119942 did not exist in that time.

Comment (RossComputerGuy) [reply pkgs/by-name/me/mednaffe/package.nix:12]: Alright, I understand. 👍

File: pkgs/by-name/me/mednafen-server/package.nix (hunk 3 above)
Comment (AndersonTorres) [reply pkgs/by-name/me/mednafen-server/package.nix:22]: Until a nasty `with` scope invasion appears:


NixOS/nixpkgs#292468 (comment)

File: pkgs/by-name/me/mednafen/package.nix [hunk 7]
```diff
@@ -1,33 +1,39 @@
...
 }:
 
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: {
   pname = "mednafen";
   version = "1.29.0";
 
   src = fetchurl {
-    url = "https://mednafen.github.io/releases/files/${pname}-${version}.tar.xz";
+    url = "https://mednafen.github.io/releases/files/mednafen-${finalAttrs.version}.tar.xz";
```
[the remark this answers is not in the document]
Comment (AndersonTorres) [reply pkgs/by-name/me/mednafen/package.nix:27]: https://github.com/NixOS/nixpkgs/issues/277994

**Review by RossComputerGuy — APPROVED** [review]
